File: Visual Basic .NET/LiveUpdating/MarketSimulator.vb
'
' * Xceed DataGrid for WPF - SAMPLE CODE - Live Updating Sample Application
' * Copyright (c) 2007-2024 Xceed Software Inc.
' * 
' * [MarketSimulator.vb]
' * 
' * This class provides simulated data to the Live Updating sample.
' * 
' * This file is part of the Xceed DataGrid for WPF product. The use 
' * and distribution of this Sample Code is subject to the terms 
' * and conditions referring to "Sample Code" that are specified in 
' * the XCEED SOFTWARE LICENSE AGREEMENT accompanying this product.
' *
' 


Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text

Namespace Xceed.Wpf.DataGrid.Samples.LiveUpdating
  Friend NotInheritable Class MarketSimulator
    Private Sub New()
    End Sub
    Private Const MaxValue As Double = 999.99
    Private Const MinValue As Double = 1

    Private Const MaxChangePercentageErratic As Double = 0.8
    Private Const MaxChangePercentageStable As Double = 0.01
    Private Const MaxChangePercentageSteadyGainer As Double = 0.03
    Private Const MaxChangePercentageSteadyLooser As Double = 0.03

    Private Const Letters As String = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"

    Private Shared Randomizer As New Random()

    Private Shared Function NextInteger() As Integer

      SyncLock MarketSimulator.Randomizer
        Return MarketSimulator.Randomizer.Next()
      End SyncLock

    End Function

    Private Shared Function NextInteger(ByVal minValue As Integer, ByVal maxValue As Integer) As Integer

      SyncLock MarketSimulator.Randomizer
        Return MarketSimulator.Randomizer.Next(minValue, maxValue)
      End SyncLock

    End Function

    Private Shared Function NextDouble() As Double

      SyncLock MarketSimulator.Randomizer
        Return MarketSimulator.Randomizer.NextDouble()
      End SyncLock

    End Function

    Friend Shared Function GenerateRandomSymbol() As String
      Dim stringBuilder As New StringBuilder()
      For i As Integer = 0 To MarketSimulator.NextInteger(3, 5) - 1

        stringBuilder.Append(MarketSimulator.Letters(MarketSimulator.NextInteger(0, Letters.Length)))
      Next

      Return stringBuilder.ToString()
    End Function

    Friend Shared Function GenerateQuoteInfo(ByVal symbol As String) As QuoteInfo
      Dim popularity As Popularities
      Dim tendency As Tendencies

      If [String].Equals(symbol, "XCEED") Then
        popularity = Popularities.Active
        tendency = Tendencies.SteadyGainer
      Else
        popularity = DirectCast(MarketSimulator.NextInteger(0, [Enum].GetValues(GetType(Popularities)).Length), Popularities)
        tendency = DirectCast(MarketSimulator.NextInteger(0, [Enum].GetValues(GetType(Tendencies)).Length), Tendencies)
      End If

      Dim previousClose As Double = MarketSimulator.NextInteger(5, 150)
      Dim openingValue As Double = MarketSimulator.SimulateChange(previousClose, popularity, tendency)
      Dim lastTrade As Double = MarketSimulator.SimulateChange(openingValue, popularity, tendency)

      Return New QuoteInfo(symbol, previousClose, openingValue, lastTrade, popularity, tendency)
    End Function

    Public Shared Function SimulateChange(ByVal quoteInfo As QuoteInfo) As Double
      Return MarketSimulator.SimulateChange(quoteInfo.LastTrade, quoteInfo.Popularity, quoteInfo.Tendency)
    End Function

    Private Shared Function SimulateChange(ByVal currentValue As Double, ByVal popularity As Popularities, ByVal tendency As Tendencies) As Double
      Dim newValue As Double = currentValue

      Dim activityFactor As Double = (MarketSimulator.NextDouble() * 100)

      If popularity = Popularities.Passive Then
        If activityFactor < 66 Then
          Return newValue
        End If
      ElseIf activityFactor < 33 Then
        Return newValue
      End If

      If (tendency = Tendencies.Stable) AndAlso (activityFactor < 33) Then
        Return newValue
      End If


      Dim changeFactor As Double = (MarketSimulator.NextDouble() * 100)

      Select Case tendency
        Case Tendencies.Erratic
          Dim change As Double = currentValue * (MarketSimulator.NextDouble() * MarketSimulator.MaxChangePercentageErratic)

          If MarketSimulator.NextDouble() < 0.5 Then
            newValue = (currentValue - change)
          Else
            newValue = (currentValue + change)
          End If

          Exit Select
        Case Tendencies.SteadyGainer

          Dim change As Double = currentValue * (MarketSimulator.NextDouble() * MarketSimulator.MaxChangePercentageSteadyGainer)

          If MarketSimulator.NextDouble() < 0.25 Then
            newValue = (currentValue - change)
          Else
            newValue = (currentValue + change)
          End If

          Exit Select
        Case Tendencies.SteadyLoser

          Dim change As Double = currentValue * (MarketSimulator.NextDouble() * MarketSimulator.MaxChangePercentageSteadyLooser)

          If MarketSimulator.NextDouble() < 0.75 Then
            newValue = (currentValue - change)
          Else
            newValue = (currentValue + change)
          End If

          Exit Select
        Case Else

          Dim change As Double = currentValue * (MarketSimulator.NextDouble() * MarketSimulator.MaxChangePercentageStable)

          If MarketSimulator.NextDouble() < 0.5 Then
            newValue = (currentValue - change)
          Else
            newValue = (currentValue + change)
          End If

          Exit Select
      End Select

      Return Math.Min(Math.Max(MarketSimulator.MinValue, newValue), MarketSimulator.MaxValue)
    End Function

#Region "NESTED ENUMS"

    Public Enum Tendencies
      Stable = 0
      Erratic = 1
      SteadyGainer = 2
      SteadyLoser = 3
    End Enum

    Public Enum Popularities
      Passive = 0
      Active = 1
    End Enum

#End Region
  End Class
End Namespace

File: CSharp/LiveUpdating/MarketSimulator.cs
/*
 * Xceed DataGrid for WPF - SAMPLE CODE - Live Updating Sample Application
 * Copyright (c) 2007-2024 Xceed Software Inc.
 * 
 * [MarketSimulator.cs]
 * 
 * This class provides simulated data to the Live Updating sample.
 * 
 * This file is part of the Xceed DataGrid for WPF product. The use 
 * and distribution of this Sample Code is subject to the terms 
 * and conditions referring to "Sample Code" that are specified in 
 * the XCEED SOFTWARE LICENSE AGREEMENT accompanying this product.
 *
 */

using System;
using System.Text;

namespace Xceed.Wpf.DataGrid.Samples.LiveUpdating
{
  internal static class MarketSimulator
  {
    private const double MaxValue = 999.99d;
    private const double MinValue = 1.00d;

    private const double MaxChangePercentageErratic = 0.8d;
    private const double MaxChangePercentageStable = 0.01d;
    private const double MaxChangePercentageSteadyGainer = 0.03d;
    private const double MaxChangePercentageSteadyLooser = 0.03d;

    private const string Letters = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";

    private static Random Randomizer = new Random();

    private static int Next()
    {
      lock( MarketSimulator.Randomizer )
      {
        return MarketSimulator.Randomizer.Next();
      }
    }

    private static int Next( int minValue, int maxValue )
    {
      lock( MarketSimulator.Randomizer )
      {
        return MarketSimulator.Randomizer.Next( minValue, maxValue );
      }
    }

    private static double NextDouble()
    {
      lock( MarketSimulator.Randomizer )
      {
        return MarketSimulator.Randomizer.NextDouble();
      }
    }

    internal static string GenerateRandomSymbol()
    {
      StringBuilder stringBuilder = new StringBuilder();

      for( int i = 0; i < MarketSimulator.Randomizer.Next( 3, 5 ); i++ )
      {
        stringBuilder.Append( MarketSimulator.Letters[ MarketSimulator.Randomizer.Next( 0, Letters.Length ) ] );
      }

      return stringBuilder.ToString();
    }

    internal static QuoteInfo GenerateQuoteInfo( string symbol )
    {
      Popularities popularity;
      Tendencies tendency;

      if( String.Equals( symbol, "XCEED" ) )
      {
        popularity = Popularities.Active;
        tendency = Tendencies.SteadyGainer;
      }
      else
      {
        popularity = ( Popularities )MarketSimulator.Randomizer.Next( 0, Enum.GetValues( typeof( Popularities ) ).Length );
        tendency = ( Tendencies )MarketSimulator.Randomizer.Next( 0, Enum.GetValues( typeof( Tendencies ) ).Length );
      }

      double previousClose = MarketSimulator.Randomizer.Next( 5, 150 );
      double openingValue = MarketSimulator.SimulateChange( previousClose, popularity, tendency );
      double lastTrade = MarketSimulator.SimulateChange( openingValue, popularity, tendency );

      return new QuoteInfo(
        symbol,
        previousClose,
        openingValue,
        lastTrade,
        popularity,
        tendency );
    }

    public static double SimulateChange( QuoteInfo quoteInfo )
    {
      return MarketSimulator.SimulateChange( quoteInfo.LastTrade, quoteInfo.Popularity, quoteInfo.Tendency );
    }

    private static double SimulateChange( double currentValue, Popularities popularity, Tendencies tendency )
    {
      double newValue = currentValue;

      double activityFactor = ( MarketSimulator.NextDouble() * 100 );

      if( popularity == Popularities.Passive )
      {
        if( activityFactor < 66 )
          return newValue;
      }
      else if( activityFactor < 33 )
      {
        return newValue;
      }

      if( ( tendency == Tendencies.Stable ) && ( activityFactor < 33 ) )
        return newValue;


      double changeFactor = ( MarketSimulator.NextDouble() * 100 );

      switch( tendency )
      {
        case Tendencies.Erratic:
          {
            double change = currentValue * ( MarketSimulator.NextDouble() * MarketSimulator.MaxChangePercentageErratic );

            if( MarketSimulator.NextDouble() < 0.5 )
            {
              newValue = ( currentValue - change );
            }
            else
            {
              newValue = ( currentValue + change );
            }

            break;
          }

        case Tendencies.SteadyGainer:
          {
            double change = currentValue * ( MarketSimulator.NextDouble() * MarketSimulator.MaxChangePercentageSteadyGainer );

            if( MarketSimulator.NextDouble() < 0.25 )
            {
              newValue = ( currentValue - change );
            }
            else
            {
              newValue = ( currentValue + change );
            }

            break;
          }

        case Tendencies.SteadyLoser:
          {
            double change = currentValue * ( MarketSimulator.NextDouble() * MarketSimulator.MaxChangePercentageSteadyLooser );

            if( MarketSimulator.NextDouble() < 0.75 )
            {
              newValue = ( currentValue - change );
            }
            else
            {
              newValue = ( currentValue + change );
            }

            break;
          }

        default:
          {
            double change = currentValue * ( MarketSimulator.NextDouble() * MarketSimulator.MaxChangePercentageStable );

            if( MarketSimulator.NextDouble() < 0.5 )
            {
              newValue = ( currentValue - change );
            }
            else
            {
              newValue = ( currentValue + change );
            }

            break;
          }
      }

      return Math.Min(
        Math.Max( MarketSimulator.MinValue, newValue ),
        MarketSimulator.MaxValue );
    }

    #region NESTED ENUMS

    public enum Tendencies
    {
      Stable = 0,
      Erratic = 1,
      SteadyGainer = 2,
      SteadyLoser = 3
    }

    public enum Popularities
    {
      Passive = 0,
      Active = 1
    }

    #endregion NESTED ENUMS
  }
}
